File: sdk-typescript/arcflow/hitl.ts
/** Human-in-the-loop configuration and errors. */

import { ArcFlowError } from "./exceptions.js";

export interface HitlConfigOptions {
  approvalKey: string;
  timeoutSeconds?: number;
  interrupt?: boolean;
}

export class HitlConfig {
  readonly approvalKey: string;
  readonly timeoutSeconds: number;
  readonly interrupt: boolean;

  constructor(options: HitlConfigOptions | string) {
    if (typeof options === "string") {
      this.approvalKey = options;
      this.timeoutSeconds = 3600;
      this.interrupt = true;
      return;
    }
    const key = options.approvalKey.trim();
    if (!key) {
      throw new Error("[ArcFlow] HitlConfig requires a non-empty approvalKey.");
    }
    this.approvalKey = key;
    this.timeoutSeconds = options.timeoutSeconds ?? 3600;
    this.interrupt = options.interrupt ?? true;
  }

  toJson(): string {
    return JSON.stringify({
      approval_key: this.approvalKey,
      timeout_seconds: this.timeoutSeconds,
      interrupt: this.interrupt,
    });
  }
}

export class HumanRejectedError extends ArcFlowError {
  readonly approvalKey: string | undefined;

  constructor(message: string, approvalKey?: string) {
    super(message);
    this.name = "HumanRejectedError";
    this.approvalKey = approvalKey;
  }
}

export class WorkflowInterruptedError extends ArcFlowError {
  readonly runId: string;
  readonly approvalKey: string;
  readonly expiresAt: string | undefined;

  constructor(
    message: string,
    runId: string,
    approvalKey: string,
    expiresAt?: string,
  ) {
    super(message);
    this.name = "WorkflowInterruptedError";
    this.runId = runId;
    this.approvalKey = approvalKey;
    this.expiresAt = expiresAt;
  }
}

File: sdk-python/arcflow/hitl.py
"""Human-in-the-loop configuration and errors."""

from __future__ import annotations

from dataclasses import dataclass

from arcflow.exceptions import ArcFlowError


@dataclass(frozen=True)
class HitlConfig:
    """Declarative human approval gate on a workflow step."""

    approval_key: str
    timeout_seconds: int = 3600
    interrupt: bool = True

    def to_json(self) -> str:
        import json

        return json.dumps(
            {
                "approval_key": self.approval_key,
                "timeout_seconds": self.timeout_seconds,
                "interrupt": self.interrupt,
            }
        )


class HumanRejectedError(ArcFlowError):
    """Raised when a human rejects an approval request."""

    def __init__(self, message: str, *, approval_key: str | None = None) -> None:
        super().__init__(message)
        self.approval_key = approval_key


class WorkflowInterruptedError(ArcFlowError):
    """Raised when a workflow pauses for human approval."""

    def __init__(
        self,
        message: str,
        *,
        run_id: str,
        approval_key: str,
        expires_at: str | None = None,
    ) -> None:
        super().__init__(message)
        self.run_id = run_id
        self.approval_key = approval_key
        self.expires_at = expires_at
